Application Mixins (#87)
buildpacks/rfcs: text/0000-app-mixins.md

- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

This is a proposal for allowing application developers (buildpack users) to specify mixins that will be dynamically included with their build and run images.

# Motivation
[motivation]: #motivation

Mixins already allow buildpack authors to create buildpacks that depend on an extended set of OS packages without affecting build time, but it's common for application code to depend on OS packages too. But allowing buildpacks to arbitrarily install OS packages during build would drastically increase build time, especially when CNB tooling is used to build or rebase many apps with similar package requirements. For that reason, this proposal defines a mechanism to implement mixin support for application developers without significant impact on build performance.

# What it is
[what-it-is]: #what-it-is

- *application developer* - a person who is using buildpacks to transform their application code into an OCI image
- *stackpack* - a new type of buildpack that runs against the stack image(s) instead of an app

> A application developer may specify a list of mixins in their application's `project.toml` file like this:

Reviewer: **An** application developer

Author: Fixed in 9f39572

* Can only create one layer
* Does not have access to the application source code
* Is run before all regular buildpacks
* Is run against both the build and run images
* Is distributed with the builder image
* May not write to the `/layers` directory

The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.

For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.

A stackpack is included in a builder by defining it in the `builder.toml` in the `[[stack.buildpacks]]` array of tables:

```
[[stack.buildpacks]]
id = "<stackpack id>"
uri = "<uri to stackpack>"

> ```

Reviewer 2: I feel like this has been a persistent point of confusion for me. In working group I thought the consensus was that `stackpacks` would ship on the stack images (build and run images) rather than on the builder image.

This would allow us to avoid having to output two images from `create-builder`. We could exclude the stackpacks from the final app image either by adding metadata to the run-image that will allow us to selectively exclude these layers or by adding a layer that [whitesout](https://github.com/opencontainers/image-spec/blob/master/layer.md#whiteouts) the buildpack dir

Author: Yea, that's right. I'm wrong in the text of the RFC. Will update

Author: @ekcasey do you think the stack should contain the buildpack before the builder image is created, or can the `create-builder` command inject the stackpack(s) (outputting new build and run images w/ the stackpack(s)).

Reviewer 2: I think the build and run images should ship with the stackpacks pre-installed.

Author: @ekcasey should i change `uri` to `path`?

Reviewer 2: @jkutner Do we even need to list `stack.buildpacks` if they are already in the image? The `[[buildpacks]]` section typically describe where to fetch buildpacks that should be added to the image.

Is the point to enable/disable them (I wonder if default to enabled is the right call)? In any case I think we only need id/version.

Reviewer 2: I am imagining that stackpacks live either in `/cnb/buildpacks/<id>/<version>` or in an analagous `/cnb/stackpacks`, and can therefore be looked up by `id` and `version`.

Author: Addressed in 1881dd3dead0a35b3335c99cfad893d1e2ba24b1

> ```

Reviewer 2: I feel like this has been a persistent point of confusion for me. In working group I thought the consensus was that `stackpacks` would ship on the stack images (build and run images) rather than on the builder image.

This would allow us to avoid having to output two images from `create-builder`. We could exclude the stackpacks from the final app image either by adding metadata to the run-image that will allow us to selectively exclude these layers or by adding a layer that [whitesout](https://github.com/opencontainers/image-spec/blob/master/layer.md#whiteouts) the buildpack dir

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

- How should a buildpack be identified as a stackpack?
    - Not everyone likes `privileged`
    - a special `key` in the `buildpack.toml`?
    - a `stackpack.toml`?
- Should the stackpack's detect have read-only access to the app?
    - In this proposal it does, and in the future a more generic "root buildpack" construct would need it too.

# Spec. Changes (OPTIONAL)
[spec-changes]: #spec-changes

## Stackpacks

Stackpacks are identical to other buildpacks, with the following exceptions:

1. The `<layers>` directory is NOT writable
1. The working directory WILL NOT contain application source code.
1. All changes made to the filesystem (with the exception of `/tmp`) during the execution of the stackpack's `bin/build` will be snaphotted and stored as a single layer.
1. A `launch.toml` WILL NOT be honored.

## launch.toml (TOML)

```
[[excludes]]
paths = ["<sub-path glob>"]
cache = false
```

Where:

* `paths` = a list of paths to exclude from the layer

> * `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.

Reviewer 2: Doesn't this conflict with line 175

> A `launch.toml` WILL NOT be honored.

Sorry, I know this is still a draft, just trying to get a sense of it.

Author: Addressed in 9f39572

# Meta
[meta]: #meta
- Name: Application Mixins
- Start Date: 2020-06-22
- Author(s): [Joe Kutner](https://github.com/jkutner/)
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

This is a proposal for allowing application developers (buildpack users) to specify mixins that will be dynamically included with their build and run images.

# Motivation
[motivation]: #motivation

> Mixins already allow buildpack authors to create buildpacks that depend on an extended set of OS packages without affecting build time, but it's common for application code to depend on OS packages too. But allowing buildpacks to arbitrarily install OS packages during build would drastically increase build time, especially when CNB tooling is used to build or rebase many apps with similar package requirements. For that reason, this proposal defines a mechanism to implement mixin support for application developers without significant impact on build performance.

Reviewer 3: ```suggestion
Mixins already allow buildpack authors to create buildpacks that depend on an extended set of OS packages without affecting build time, but it's common for application code to depend on OS packages too. But allowing buildpacks to arbitrarily install OS packages during build would drastically increase build time, especially when CNB tooling is used to build or rebase many apps with similar package requirements.

For that reason, this proposal defines a mechanism to implement dynamic installation of mixins at build time while minimizing the performance impact. This is accomplished by allowing stack images to satisfy mixin requires statically and/or reject dynamic mixin installation early in the build process.
```

# What it is
[what-it-is]: #what-it-is

- *application developer* - a person who is using buildpacks to transform their application code into an OCI image
- *root buildpack* - a new type of buildpack that runs as the root user
- *stackpack* - a type of root buildpack that runs against the stack image(s) instead of an app

An application developer may specify a list of mixins in their application's `project.toml` file like this:

```toml
[build]
mixins = [ "<mixin name>" ]
```

When a command like `pack build` is run, the list of mixins will be processed before buildpacks are run. For each mixin name, the following will happen:

* If the mixin name is prefixed with `build:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the build image.
* If the mixin name is prefixed with `run:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the run image.
* If the mixin name does not have a prefix it will be dynamically added to both the build and run stack images.

# How it Works
[how-it-works]: #how-it-works

> When a list of mixins are defined in a `project.toml` and `pack build` is run against that project, the following will happen:

Reviewer 3: I thought they're supposed to come from the build plan (contributed by other buildpacks)? 
```suggestion
When a list of mixins are required by buildpacks via the build plan the build phase starts:
```

1. The lifecycle will compare the list of mixins to those provided by the stack. If all mixin names are provided by the stack, no further action is required.

> 1. In any requested mixin is not provided by the stack, the lifecycle will run the detect phase for all stackpacks defined in the builder.

Reviewer 4: ```suggestion
1. If any requested mixin is not provided by the stack, the lifecycle will run the detect phase for all stackpacks defined in the builder.
```

1. The lifecycle will compare the mixins added to the build plan to see if they match they required mixins.

> 1. If at least one stackpack passes detect and provides the required mixin(s), the lifecycle will execute the stackpack build phase for passing stackpack(s). If no stackpacks pass detect, or no stackpacks provide the required mixin, the build will fail with a message that describes the mixins that could not be provided.

Reviewer 4: ```suggestion
1. If at least one stackpack passes detect and provides the required mixin(s), the lifecycle will execute the stackpack build phase for all passing stackpack(s). If no stackpacks pass detect, or no stackpacks provide the required mixin, the build will fail with a message that describes the mixins that could not be provided.
```

1. During the lifecycle's build phase, the stackpacks that passed detection will run against the build and run images accordingly (see details below). All stackpacks will be run before the user's buildpacks.

## Stackpacks

 A stackpacks is a special case of buildpack that has the following properties:

* Is run as the `root` user
* Configured with `privileged = true` in the `buildpack.toml`
* Can only create one layer
* Does not have access to the application source code
* Is run before all regular buildpacks
* Is run against both the build and run images

> * Is distributed with the builder image

Reviewer 3: ```suggestion
* Is distributed with the stack run and/or build image
```

* May not write to the `/layers` directory

The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.

For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.

A stackpack is included in a builder by defining it in the `builder.toml` in the `[[stack.buildpacks]]` array of tables:

```
[[stack.buildpacks]]
id = "<stackpack id>"
uri = "<uri to stackpack>"
```

> Each stackpack included in the builder will execute until all of a project's mixins have been provided.

Reviewer 4: This wording to me reads as ~"once a project's mixins have been provided by an earlier stackpack in the execution order, remaining stackpacks won't be executed".

However in the "how it works" section, the wording is:

> If at least one stackpack passes detect and provides the required mixin(s), the lifecycle will execute the stackpack build phase for passing stackpack(s).

...which suggests that instead there is no "early return" and all passing stackpacks will be run, even if an earlier one in the sequence provides the required mixin?

The stackpack's snapshot layer may be modified by writing a `launch.toml` file. The `[[processes]]` and `[[slices]]` tables may be used as normal, and a new `[[excludes]]` table will be introduced with the following schema:

```
[[excludes]]
paths = ["<sub-path glob>"]
cache = false
```

Where:

* `paths` = a list of paths to exclude from the layer
* `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.

## Rebasing an App

Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. Then, the rebase operation can be performed as normal, while including the stackpack layers as part of the stack. This will be made possible by including the stackpack in the run-image, but because the stackpack detect phase is not run, the operation does not need access to the application source.

## Example: Apt Buildpack

> A buildpack that installs a custom certificate would have a `buildpack.toml` like this:

Reviewer 4: Is the description here wrong? It seems to be describing the second example (custom certificate) rather than the APT buildpack example?

I think it might also be helpful to include an example `project.toml` of how one might use this APT buildpack, similar to the `project.toml` example for the CA certificate case?

Author: whoops. fixed in efb1fa6

 ```toml
[buildpack]
id = "example/apt"
privileged = true

> mixins = ["[^=]+"]

Reviewer 4: This `mixins` pattern seems quite broad - won't it mean this buildpack matches against too many mixin names? (It's quite possible I'm misunderstanding how the matching works here)

Author: The mixins concept is intentionally flexible as a whole. But it can be more narrowly scoped for each stack. For example, [a mixin is defined as a "package" for the bionic stack](https://github.com/buildpacks/rfcs/blob/main/text/0031-bionic-mixins.md). Because the stackpack in this example only works with the bionic stack, it can safely assume the mixin is a package.

Author: Addressed in 8cd657fc613769f29f84ca88673415504d57ea4c

 ```bash
#!/usr/bin/env bash

apt update -y

for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
  apt install $package
done

cat << EOF > launch.toml
[[excludes]]
paths = [ "/var/cache" ]
cache = true
EOF
```

## Example: CA Certificate Buildpack

A buildpack that works on the `ubuntu-20` stack, which defines a `type=` prefix for mixins, and installs customer CA Certificates would have a `buildpack.toml` that looks like this:

```toml
[buildpack]
id = "example/cacerts"
privileged = true
mixins = ["type=cacert"]

[[stacks]]
id = "ubuntu-20"
```

Its `bin/detect` would have the following contents:

 ```bash
#!/usr/bin/env bash

exit 0
```

Its `bin/build` would have the following contents:

 ```bash
#!/usr/bin/env bash

for cert_path in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
  cert_file=$(basename $cert_path)
  cp $cert_path /usr/share/ca-certificates/$cert_file
  chmod 644 /usr/share/ca-certificates/$cert_file
done

update-ca-certificates
```

An application that is using this stackpack would the following `project.toml`:

```toml
[build]
mixins = [ "type=cacert:mycerts/database.crt" ]
```

# Drawbacks
[drawbacks]: #drawbacks

> - Mixins are less flexible that a generic [root buildpack](https://github.com/buildpacks/rfcs/blob/app-image-ext-buildpacks/text/0000-app-image-extensions.md).

Reviewer 4: ```suggestion
- Mixins are less flexible than a generic [root buildpack](https://github.com/buildpacks/rfcs/blob/app-image-ext-buildpacks/text/0000-app-image-extensions.md).
```

# Alternatives
[alternatives]: #alternatives

- [Add root buildpack as interface for app image extension RFC](https://github.com/buildpacks/rfcs/pull/77)
- [App Image Extensions (OS packages)](https://github.com/buildpacks/rfcs/pull/23)

> - [Root Buildpacks](https://github.com/buildpacks/rfcs/blob/root-buildpacks/text/0000-root-buildpacks.md): Allow application developers to use root buildpacks in `project.toml`. This would have significant performance implications, and creates a "foot gun" in qhich end users could build images they are not able to maintain. For this reason, we are holding off on a generic root buildpack feature.

Reviewer 4: ```suggestion
- [Root Buildpacks](https://github.com/buildpacks/rfcs/blob/root-buildpacks/text/0000-root-buildpacks.md): Allow application developers to use root buildpacks in `project.toml`. This would have significant performance implications, and creates a "foot gun" in which end users could build images they are not able to maintain. For this reason, we are holding off on a generic root buildpack feature.
```

# Prior Art
[prior-art]: #prior-art

- [RFC-0006: Stage specific mixins](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)
- The term "non-idempotent" is used in section 9.1.2 of [Hypertext Transfer Protocol -- HTTP/1.1 (RFC 2616)](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html)

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

- Should the stackpack's detect have read-only access to the app?
    - Does a stackpack even need a detect phase?
    - This would likely be driven by a stackpack that does not provide mixins, but instead dynamically contributes to the build plan based on the contents of the app source code. I don't know if we have a use case for this, but I can imaging a buildpack that reads environment variables as input to some function.

# Spec. Changes (OPTIONAL)
[spec-changes]: #spec-changes

## Stackpacks

Stackpacks are identical to other buildpacks, with the following exceptions:

> 1. The `<layers>` directory is NOT writable

Reviewer 4: ```suggestion
1. The `<layers>` directory is NOT writable.
```

1. The working directory WILL NOT contain application source code during the build phase.

> 1. All changes made to the filesystem (with the exception of `/tmp`) during the execution of the stackpack's `bin/build` will be snaphotted and stored as a single layer.

Reviewer 4: ```suggestion
1. All changes made to the filesystem (with the exception of `/tmp`) during the execution of the stackpack's `bin/build` will be snapshotted and stored as a single layer.
```

For that reason, this proposal defines a mechanism to implement dynamic installation of mixins at build time while minimizing the performance impact. This is accomplished by allowing stack images to satisfy mixin requires statically and/or reject dynamic mixin installation early in the build process.

# What it is
[what-it-is]: #what-it-is

- *application developer* - a person who is using buildpacks to transform their application code into an OCI image
- *root buildpack* - a new type of buildpack that runs as the root user
- *stackpack* - a type of root buildpack that runs against the stack image(s) instead of an app. It is distinguished by a static list of mixins it can provide.

An application developer may specify a list of mixins in their application's `project.toml` file like this:

```toml
[build]
mixins = [ "<mixin name>" ]
```

When a command like `pack build` is run, the list of mixins will be processed before buildpacks are run. For each mixin name, the following will happen:

> * If the mixin name is prefixed with `build:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the build image.

Reviewer 5: Newbie question: I'm assuming this happens prior to any phase of the buildpack lifecycle happening. As in, it happens when we `pack create-builder ...`. Which means all the buildpack lifecycle steps will never read to read/write from/to `/`. Is this accurate?

Reviewer 5: (By lifecycle steps I mean: detector, analyzer, restorer, builder, exporter)

Author: No, the stackpacks would not execute during `create-builder`. They run during the normal lifecycle phases, but in as the `root` user, which allows them to write to `/`). Does that help?

Reviewer 5: I see. So in the line above you mean the mixin will be dynamically added to the builder* image when prefixed with `build:`. 

I'm curious, if implemented, which lifecycle phase will need to read/write to/from `/`. All of them or just the `builder` phase?

Author: Only the build phase will need (or more correctly "be able") to write to `/`

 ```bash
#!/usr/bin/env bash

exit 0
```

Its `bin/build` would have the following contents:

 ```bash
#!/usr/bin/env bash

apt update -y

for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
  apt install $package
done

cat << EOF > launch.toml
[[excludes]]
paths = [ "/var/cache" ]
cache = true
EOF
```

## Example: CA Certificate Buildpack

A buildpack that works on the `ubuntu-20` stack, which defines a `type=` prefix for mixins, and installs custom CA Certificates would have a `buildpack.toml` that looks like this:

```toml
[buildpack]
id = "example/cacerts"
privileged = true
mixins = ["type=cacert"]

[[stacks]]
id = "ubuntu-20"
```

Its `bin/detect` would have the following contents:

 ```bash
#!/usr/bin/env bash

exit 0
```

Its `bin/build` would have the following contents:

 ```bash
#!/usr/bin/env bash

for cert_path in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
  cert_file=$(basename $cert_path)
  cp $cert_path /usr/share/ca-certificates/$cert_file
  chmod 644 /usr/share/ca-certificates/$cert_file
done

update-ca-certificates
```

An application that is using this stackpack would the following `project.toml`:

```toml
[build]

> mixins = [ "type=cacert:mycerts/database.crt" ]

Reviewer 6: What do you think about using `,` instead of `:`
```suggestion
mixins = [ "type=cacert,mycerts/database.crt" ]
```

Sort of like docker
`docker run --read-only --mount type=volume,target=/icanwrite busybox`

Author: This has changed in 6a3cfba

An application developer may specify a list of mixins in their application's `project.toml` file like this:

```toml
[build]
mixins = [ "<mixin name>" ]
```

When a command like `pack build` is run, the list of mixins will be processed before buildpacks are run. For each mixin name, the following will happen:

* If the mixin name is prefixed with `build:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the build image.
* If the mixin name is prefixed with `run:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the run image.
* If the mixin name does not have a prefix it will be dynamically added to both the build and run stack images.

# How it Works
[how-it-works]: #how-it-works

When a list of mixins are required by buildpacks via the build plan and the build phase starts:

1. The lifecycle will compare the list of mixins to those provided by the stack. If all mixin names are provided by the stack, no further action is required.

> 1. If any requested mixin is not provided by the stack, the lifecycle will run the detect phase for all stackpacks defined in the builder.

Reviewer 3: Can we run `/bin/detect` in parallel with other buidpacks' `/bin/detect`? Then the stackpack could opt-out entirely (be incapable of providing mixins) if its detect fails.

(Should stackpacks have an optional flag and fail all builds if they fail detection?)

1. If at least one stackpack passes detect and provides the required mixin(s), the lifecycle will execute the stackpack build phase for all passing stackpack(s). If no stackpacks pass detect, or no stackpacks provide the required mixin, the build will fail with a message that describes the mixins that could not be provided.
1. During the lifecycle's build phase, the stackpacks that passed detection will run against the build and run images accordingly (see details below). All stackpacks will be run before the user's buildpacks.

## Stackpacks

 A stackpacks is a special case of buildpack that has the following properties:

* Is run as the `root` user
* Configured with `privileged = true` in the `buildpack.toml`
* Can only create one layer
* Does not have access to the application source code
* Is run before all regular buildpacks
* Is run against both the build and run images
* Is distributed with the stack run and/or build image
* May not write to the `/layers` directory

The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.

> For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.

Reviewer 7: This feature would be super valuable for Windows images for all the same reasons, perhaps more-so since Windows artifacts are often more strict about their location. 

Kaniko appears to have no plans to support Windows containers (https://github.com/GoogleContainerTools/kaniko/issues/111), but does this spec language leave open the option for lifecycle to use alternate implementations? Is that the intent of: 
> Alternatively, a platform may store a new stack image to cache the changes?.

> A builder can use a stackpack by defining it in the `builder.toml` in the `[[stack.buildpacks]]` array of tables:

Reviewer 3: Given that the run image must also contain stackpacks, I suggest we require the stackpacks be present on the stack images themselves and not added as part of `pack create-builder`.

 ```toml
[buildpack]
id = "example/apt"
privileged = true
mixins = ["[^=]+"]

[[stacks]]
id = "io.buildpacks.stacks.bionic"
```

Its `bin/detect` would have the following contents:

 ```bash
#!/usr/bin/env bash

exit 0
```

Its `bin/build` would have the following contents:

 ```bash
#!/usr/bin/env bash

apt update -y

for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
  apt install $package
done

cat << EOF > launch.toml
[[excludes]]
paths = [ "/var/cache" ]
cache = true
EOF
```

## Example: CA Certificate Buildpack

Support for custom CA Certificates can be accomplished with two buildpacks: a stackpack that can install the cert, and a normal buildpack that can provide a cert in the build plan.

### CA Cert Installer Stackpack

A buildpack that works on the `ubuntu-20` stack, which defines a `type=` prefix for mixins, and installs custom CA Certificates would have a `buildpack.toml` that looks like this:

```toml
[buildpack]
id = "example/cacerts"
privileged = true
mixins = ["cacert"]

[[stacks]]
id = "ubuntu-20"
```

Its `bin/detect` would have the following contents:

 ```bash
#!/usr/bin/env bash

cat << EOF > $2

> [[provides]]

Reviewer 3: I support `/bin/detect` for stackpacks, so that they can validate that the stack image itself is in a valid state. However, I'm reluctant to allow mixins to be specified dynamically during detect, given that the app directory is not available.

Author: @sclevine what if it's not based on app source, but some other input from the platform (like a list of mixins that are needed)

Author: Addressed in 8ee98d430e2ce55652cd53ae5ed9f344bd888889

Reviewer 2: > However, I'm reluctant to allow mixins to be specified dynamically during detect, given that the app directory is not available.

I agree that the stackpacks shouldn't dynamically decide which mixins to provide during detect, but regular buildpacks (which have access to the app dir) should probably be able to require specific mixins during detect.

- Supersedes: N/A

# Summary
[summary]: #summary

This is a proposal for allowing application developers (buildpack users) to specify mixins that will be dynamically included with their build and run images.

# Motivation
[motivation]: #motivation

Mixins already allow buildpack authors to create buildpacks that depend on an extended set of OS packages without affecting build time, but it's common for application code to depend on OS packages too. But allowing buildpacks to arbitrarily install OS packages during build would drastically increase build time, especially when CNB tooling is used to build or rebase many apps with similar package requirements.

For that reason, this proposal defines a mechanism to implement dynamic installation of mixins at build time while minimizing the performance impact. This is accomplished by allowing stack images to satisfy mixin requires statically and/or reject dynamic mixin installation early in the build process.

# What it is
[what-it-is]: #what-it-is

- *application developer* - a person who is using buildpacks to transform their application code into an OCI image
- *root buildpack* - a new type of buildpack that runs as the root user

> - *stackpack* - a type of root buildpack that runs against the stack image(s) instead of an app. It is distinguished by a static list of mixins it can provide.

Reviewer 8: ```suggestion
- *stack buildpack* - a type of root buildpack that runs against the stack image(s) instead of an app. It is distinguished by a static list of mixins it can provide.
```

I propose keeping the naming more inline with "buildpacks" and simply prefix the type as the name as a whole. Specially, if we can envision being able to provide them at build time from a platform's perspective.

## Stackpacks

 A stackpacks is a special case of buildpack that has the following properties:

* Is run as the `root` user
* Configured with `privileged = true` in the `buildpack.toml`
* Can only create one layer
* Does not have access to the application source code
* Is run before all regular buildpacks
* Is run against both the build and run images
* Is distributed with the stack run and/or build image
* May not write to the `/layers` directory

The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.

For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.

> A stack can provide stackpacks by including them in the `/cnb/stackpacks` directory. By default, an implicit order will be created for all stackpacks included in a stack. The order can be overridden in the `builder.toml` with the following configuration:

Reviewer 8: ```suggestion
A stack can provide stackpacks by including them in the `/cnb/stack/buildpacks` directory. By default, an implicit order will be created for all stackpacks included in a stack. The order can be overridden in the `builder.toml` with the following configuration:
```

## Example: Apt Buildpack

A buildpack that can install an arbitrary list of mixins would have a `buildpack.toml` like this:

 ```toml
[buildpack]
id = "example/apt"
privileged = true
provides-mixins = true

[[stacks]]
id = "io.buildpacks.stacks.bionic"
```

Its `bin/detect` would have the following contents:

 ```bash
#!/usr/bin/env bash

# TODO read the mixins.toml (which has same schema as build plan)
for package in $(cat mixins.toml | yj -t | jq -r ".requires | .[] | .name"); do
  cat << EOF >> $2
[[provides]]
name = "${package}"
EOF
done

exit 0
```

Its `bin/build` would have the following contents:

 ```bash
#!/usr/bin/env bash

apt update -y

for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
  apt install $package
done

cat << EOF > launch.toml
[[excludes]]
paths = [ "/var/cache" ]
cache = true
EOF
```

## Example: CA Certificate Buildpack

Support for custom CA Certificates can be accomplished with two buildpacks: a stackpack that can install the cert, and a normal buildpack that can provide a cert in the build plan.

### CA Cert Installer Stackpack

> A buildpack that works on the `ubuntu-20` stack, which defines a `type=` prefix for mixins, and installs custom CA Certificates would have a `buildpack.toml` that looks like this:

Reviewer 6: This is no longer true in the example right?

Author: Fixed in 8cd657fc613769f29f84ca88673415504d57ea4c

```toml
[buildpack]
id = "example/cacerts"
privileged = true

[[stacks]]
id = "io.buildpacks.stacks.bionic"
```

Its `bin/detect` would have the following contents:

 ```bash
#!/usr/bin/env bash

cat << EOF > $2

> [[provides]]

Reviewer 6: Is there any way for this buildpack to _only_ run on run images and not on the build image? The mixins has `build:` and `run:`, can this example only run on run images?

* Is distributed with the stack run and/or build image
* May not write to the `/layers` directory

The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.

For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase excluding the a few specific directories and files.

Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.

A stack can provide stackpacks by including them in the `/cnb/stack/buildpacks` directory. By default, an implicit order will be created for all stackpacks included in a stack. The order can be overridden in the `builder.toml` with the following configuration:

```
[[stack.order]]
  [[stack.order.group]]
    id = "<stackpack id>"
```

> A stackpack will only execute if it passes detection. When the stackpack is executed, it's detect and build scripts use the same parameters as the regular buildpacks. But the arguments may differ. FOr example, the first positional argument to the `bin/build` (the `<layers>` directory) MUST NOT be a standard layers directory adhering to the [Buildpacks Build specification](https://github.com/buildpacks/spec/blob/main/buildpack.md#build).

Reviewer 6: ```suggestion
A stackpack will only execute if it passes detection. When the stackpack is executed, it's detect and build scripts use the same parameters as the regular buildpacks. But the arguments may differ. For example, the first positional argument to the `bin/build` (the `<layers>` directory) MUST NOT be a standard layers directory adhering to the [Buildpacks Build specification](https://github.com/buildpacks/spec/blob/main/buildpack.md#build).
```
